tests/dtypes/test_dataset_to_numpy.py

```diff
@@ -0,0 +1,110 @@
+import numpy as np
+import pandas as pd
+import pytest
+
+import reciprocalspaceship as rs
+
+
+@pytest.mark.parametrize("dtype", [None, np.int32, np.float32, np.float64, object])
+@pytest.mark.parametrize("hasna", [True, False])
+def test_to_numpy_ints(dtype, hasna):
+    """
+    Test DataSet.to_numpy() with int32-backed MTZDtypes
+    """
+    ds = rs.DataSet(
+        {
+            "X": np.random.randint(0, 100, size=100),
+            "Y": np.random.randint(0, 100, size=100),
+        }
+    ).infer_mtz_dtypes()
+
+    if hasna:
+        ds.loc[0, "X"] = np.nan
+
+    # Can't represent array with np.int32
+    if hasna and dtype is np.int32:
+        with pytest.raises(ValueError):
+            arr = ds.to_numpy(dtype=dtype)
+    else:
+        arr = ds.to_numpy(dtype=dtype)
+
+        if dtype is None:
+            # Fall back to float32
+            if hasna:
+                assert arr.dtype == np.float32
+            # Fall back to int32
+            else:
+                assert arr.dtype == np.int32
+        else:
+            assert arr.dtype == dtype
+
+
+@pytest.mark.parametrize("dtype", [None, np.float32, np.float64, object])
+@pytest.mark.parametrize("hasna", [True, False])
+def test_to_numpy_floats(dtype, hasna):
+    """
+    Test DataSet.to_numpy() with float32-backed MTZDtypes
+    """
+    ds = rs.DataSet(
+        {
+            "X": np.random.random(100),
+            "Y": np.random.random(100),
+        }
+    ).infer_mtz_dtypes()
+
+    if hasna:
+        ds.loc[0, "X"] = np.nan
+
+    arr = ds.to_numpy(dtype=dtype)
+
+    if dtype is None:
+        # Fall back to float32
+        assert arr.dtype == np.float32
+    else:
+        assert arr.dtype == dtype
+
+
+@pytest.mark.parametrize("dtype", [None, np.float32, np.float64, object])
+@pytest.mark.parametrize("hasna", [True, False])
+def test_to_numpy_mtzdtypes(dtype, hasna):
```

Comment:
This test name is a little confusing, the first 3 tests all use `mtzdtypes`. Maybe `test_to_numpy_mixed_mtzdtypes` or something. Likewise maybe the previous tests should be `test_to_numpy_mtzints` and `test_to_numpy_mtzfloats`.

Reply:
I updated the test function names to be a bit more explicit